File: app/src/main/java/com/example/homestay/LoginActivity.java
package com.example.homestay;

import android.content.Intent;
import android.os.Bundle;
import android.widget.TextView;
import android.widget.Toast;
import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.lifecycle.ViewModelProvider;
import com.example.homestay.data.entity.User;
import com.example.homestay.ui.viewmodel.AuthViewModel;
import com.example.homestay.utils.AdminAuth;
import com.example.homestay.utils.SessionManager;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputEditText;
import java.util.concurrent.Executors;

public class LoginActivity extends AppCompatActivity {
  private HomestayApplication app;
  private AuthViewModel viewModel;
  private SessionManager sessionManager;

  @Override
  protected void onCreate(Bundle state) {
    super.onCreate(state);
    EdgeToEdge.enable(this);
    setContentView(R.layout.activity_login);
    ViewCompat.setOnApplyWindowInsetsListener(
        findViewById(android.R.id.content),
        (v, insets) -> {
          Insets b = insets.getInsets(WindowInsetsCompat.Type.systemBars());
          v.setPadding(b.left, b.top, b.right, b.bottom);
          return insets;
        });
    app = (HomestayApplication) getApplicationContext();
    sessionManager = new SessionManager(this);
    viewModel =
        new ViewModelProvider(this, new AuthViewModel.Factory(app.getAuthRepository(), this))
            .get(AuthViewModel.class);
    if (isAdminLoggedIn()) {
      navigateAdmin();
      return;
    }
    if (sessionManager.isLoggedIn()) {
      navigateMain();
      return;
    }
    setupViews();
    viewModel
        .getLoginResult()
        .observe(
            this,
            result -> {
              if (result == null) return;
              if (result.success && result.user != null && result.mongoUserId != null) {
                clearAdminSession();
                User user = result.user;
                sessionManager.saveSession(
                    user.getId(), result.mongoUserId, user.getEmail(), user.getFullName());
                Executors.newSingleThreadExecutor()
                    .execute(() -> app.getRepository().refreshLocalRooms());
                Toast.makeText(this, "Đăng nhập thành công!", Toast.LENGTH_SHORT).show();
                navigateMain();
              } else
                Toast.makeText(
                        this,
                        result.message == null ? "Đăng nhập thất bại" : result.message,
                        Toast.LENGTH_SHORT)
                    .show();
            });
  }

  private void setupViews() {
    TextInputEditText email = findViewById(R.id.et_email),
        password = findViewById(R.id.et_password);
    MaterialButton button = findViewById(R.id.btn_login);
    android.view.View form = findViewById(R.id.login_form_container);
    form.setAlpha(0);
    form.setTranslationY(100);
    form.animate().alpha(1).translationY(0).setDuration(800).setStartDelay(200).start();
    button.setOnClickListener(
        v -> {
          String e = text(email).trim(), p = text(password);
          if (e.isEmpty()) {
            email.setError("Vui lòng nhập email");
            return;
          }
          if (p.isEmpty()) {
            password.setError("Vui lòng nhập mật khẩu");
            return;
          }
          if (e.equalsIgnoreCase(AdminAuth.EMAIL)) {
            if (AdminAuth.authenticate(e, p)) createAdminSession();
            else Toast.makeText(this, "Email hoặc mật khẩu không đúng", Toast.LENGTH_SHORT).show();
            return;
          }
          viewModel.login(e, p);
        });
    ((TextView) findViewById(R.id.tv_register_link))
        .setOnClickListener(v -> startActivity(new Intent(this, RegisterActivity.class)));
  }

  private void createAdminSession() {
    Executors.newSingleThreadExecutor()
        .execute(
            () -> {
              User user = app.getRepository().getUserByEmail(AdminAuth.EMAIL);
              if (user == null) {
                long id =
                    app.getRepository()
                        .insertUser(
                            new User(
                                0,
                                AdminAuth.EMAIL,
                                "admin-local",
                                "AdminLocalSession@1",
                                "Administrator",
                                System.currentTimeMillis()));
                user = app.getRepository().getUserById(id);
              }
              if (user != null)
                sessionManager.saveSession(
                    user.getId(), "local-admin", user.getEmail(), user.getFullName());
              runOnUiThread(
                  () -> {
                    saveAdminSession();
                    navigateAdmin();
                  });
            });
  }

  private static String text(TextInputEditText field) {
    return field.getText() == null ? "" : field.getText().toString();
  }

  private boolean isAdminLoggedIn() {
    return getSharedPreferences("AdminSession", MODE_PRIVATE)
        .getBoolean("is_admin_logged_in", false);
  }

  private void saveAdminSession() {
    getSharedPreferences("AdminSession", MODE_PRIVATE)
        .edit()
        .putString("admin_id", "local-admin")
        .putString("admin_username", AdminAuth.EMAIL)
        .putString("admin_fullname", "Administrator")
        .putString("admin_role", "super_admin")
        .putBoolean("is_admin_logged_in", true)
        .apply();
  }

  private void clearAdminSession() {
    getSharedPreferences("AdminSession", MODE_PRIVATE).edit().clear().apply();
  }

  private void navigateMain() {
    Intent i = new Intent(this, MainActivity.class);
    i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
    startActivity(i);
    finish();
  }

  private void navigateAdmin() {
    Intent i = new Intent(this, AdminDashboardActivity.class);
    i.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
    startActivity(i);
    finish();
  }
}

File: app/src/main/java/com/example/homestay/LoginActivity.kt
package com.example.homestay

import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.lifecycleScope
import com.example.homestay.ui.viewmodel.AuthViewModel
import com.example.homestay.ui.viewmodel.AuthViewModelFactory
import com.example.homestay.utils.SessionManager
import com.google.android.material.button.MaterialButton
import com.google.android.material.textfield.TextInputEditText
import kotlinx.coroutines.launch

class LoginActivity : AppCompatActivity() {
    private val application by lazy { applicationContext as HomestayApplication }
    private val viewModel: AuthViewModel by viewModels {
        AuthViewModelFactory(application.authRepository, this)
    }
    private lateinit var sessionManager: SessionManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_login)
        
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(android.R.id.content)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        sessionManager = SessionManager(this)

        // Nếu đã đăng nhập, chuyển đến MainActivity
        if (sessionManager.isLoggedIn()) {
            navigateToMain()
            return
        }

        setupViews()
        observeLoginResult()
    }

    private fun setupViews() {
        val etEmail = findViewById<TextInputEditText>(R.id.et_email)
        val etPassword = findViewById<TextInputEditText>(R.id.et_password)
        val btnLogin = findViewById<MaterialButton>(R.id.btn_login)
        val tvRegisterLink = findViewById<android.widget.TextView>(R.id.tv_register_link)

        // Animation cho Form Container
        val loginForm = findViewById<android.view.View>(R.id.login_form_container)
        loginForm.alpha = 0f
        loginForm.translationY = 100f
        loginForm.animate()
            .alpha(1f)
            .translationY(0f)
            .setDuration(800)
            .setStartDelay(200)
            .start()

        btnLogin.setOnClickListener {
            val email = etEmail.text?.toString()?.trim() ?: ""
            val password = etPassword.text?.toString() ?: ""

            if (email.isEmpty()) {
                etEmail.error = "Vui lòng nhập email"
                return@setOnClickListener
            }

            if (password.isEmpty()) {
                etPassword.error = "Vui lòng nhập mật khẩu"
                return@setOnClickListener
            }

            if (email.equals(ADMIN_USERNAME, ignoreCase = true)) {
                if (password == ADMIN_PASSWORD) {
                    saveAdminSession()
                    navigateToAdminDashboard()
                } else {
                    Toast.makeText(this, "Tên đăng nhập hoặc mật khẩu không đúng", Toast.LENGTH_SHORT).show()
                }
                return@setOnClickListener
            }

            viewModel.login(email, password)
        }

        tvRegisterLink.setOnClickListener {
            startActivity(Intent(this, RegisterActivity::class.java))
        }
    }

    private fun observeLoginResult() {
        lifecycleScope.launch {
            viewModel.loginResult.collect { result ->
                result?.let {
                    if (it.success) {
                        val user = it.user
                        val mongoUserId = it.mongoUserId
                        if (user != null && mongoUserId != null) {
                            // Lưu session (bao gồm MongoDB ID để gọi API update)
                            sessionManager.saveSession(user.id, mongoUserId, user.email, user.fullName)
                            
                            // Sync rooms từ backend sau khi đăng nhập
                            val application = applicationContext as com.example.homestay.HomestayApplication
                            lifecycleScope.launch {
                                try {
                                    application.repository.syncRoomsFromAPI()
                                    android.util.Log.d("LoginActivity", "Rooms synced after login")
                                } catch (e: Exception) {
                                    android.util.Log.e("LoginActivity", "Error syncing rooms: ${e.message}", e)
                                }
                            }
                            
                            Toast.makeText(this@LoginActivity, "Đăng nhập thành công!", Toast.LENGTH_SHORT).show()
                            navigateToMain()
                        }
                    } else {
                        Toast.makeText(this@LoginActivity, it.message ?: "Đăng nhập thất bại", Toast.LENGTH_SHORT).show()
                    }
                }
            }
        }
    }

    private fun navigateToMain() {
        val intent = Intent(this, MainActivity::class.java)
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
        startActivity(intent)
        finish()
    }

    private fun saveAdminSession() {
        getSharedPreferences("AdminSession", MODE_PRIVATE).edit()
            .putString("admin_id", "local-admin")
            .putString("admin_username", ADMIN_USERNAME)
            .putString("admin_fullname", "Administrator")
            .putString("admin_role", "super_admin")
            .putBoolean("is_admin_logged_in", true)
            .apply()
    }

    private fun navigateToAdminDashboard() {
        val intent = Intent(this, AdminDashboardActivity::class.java)
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
        startActivity(intent)
        finish()
    }

    companion object {
        private const val ADMIN_USERNAME = "admin"
        private const val ADMIN_PASSWORD = "Admin@123"
    }
}

